### PR TITLE
ROC-5105: Paid in Full error if claimant settles before linking defendant's email

### PR DESCRIPTION
**Pull request by laurentaylor1**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ROC-5105

### Change description ###
Added check to see if defendant's email exists before sending them a paid in full email.
If defendant email doesn't exist, claimant's 'optional' defendant email is used instead, otherwise email is not sent.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandler.java [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.flywaydb.core.internal.logging.console.ConsoleLog;
 import org.springframework.context.event.EventListener;
 import org.springframework.stereotype.Component;
 import uk.gov.hmcts.cmc.claimstore.config.properties.notifications.NotificationsProperties;
 import uk.gov.hmcts.cmc.claimstore.services.notifications.NotificationService;
 import uk.gov.hmcts.cmc.domain.models.Claim;
+import uk.gov.hmcts.cmc.domain.models.otherparty.TheirDetails;
 
+import javax.swing.text.html.Option;
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandler.java]: Remove unused imports

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandler.java [hunk 2]
```diff
@@ -33,14 +38,25 @@ public void onPaidInFullEvent(PaidInFullEvent event) {
...
+                notificationsProperties.getTemplates().getEmail().getClaimantSaysDefendantHasPaidInFull(),
+                parameters,
+                referenceForDefendant(claim.getReferenceNumber())
+            )
         );
     }
 
+    private Optional<String> getDefendantEmail(Claim claim) {
+        if (StringUtils.isNotBlank(claim.getDefendantEmail())) {
+            return Optional.ofNullable(claim.getDefendantEmail());
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandler.java]: At this point it's guaranteed not null, so use `Optional.of(claim.getDefendantEmail())`

Comment (laurentaylor1) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandler.java]: Cheers

**Comment by laurentaylor1** [comment]
@sabahirfan There is an existing unit test called `sendNotificationsSendsNotificationsToDefendant()` in `uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java` that grabs the defendant's email in the sample claim. Is this sufficient? Cheers

**Comment by sabahirfan** [comment]
@laurentaylor1 These three possible scenarios needs to be tested:
1. If claim is linked, then use the defendant email from Claim
2. if claim is not linked, then use defendant email provided by claimant.
3. if we don't have defendant email at all, then it should not send the email.

The mentioned unit test just covers the one scenario. Two are still missing in unit tests.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java [hunk 3]
```diff
@@ -45,11 +48,12 @@ public void setUp() {
     public void sendNotificationsSendsNotificationsToDefendant() throws NotificationClientException {
 
         PaidInFullEvent event = new PaidInFullEvent(SampleClaim.getDefault());
+        Response fullDefenceResponse = SampleResponse.FullDefence.builder().build();
+        Claim claim = SampleClaim.getWithResponse(fullDefenceResponse);
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java]: The above two lines are not required.

Comment (laurentaylor1) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java]: Test fails without it as getWithResponse() is needed to initialise the defendant's email in the sample claim

Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java]: Please pass this claim object in event constructor.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java [hunk 4]
```diff
@@ -45,11 +48,12 @@ public void setUp() {
...
 
         handler.notifyDefendantForPaidInFull(event);
-
         verify(notificationService, once()).sendMail(
-            eq(event.getClaim().getDefendantEmail()),
+            eq(claim.getDefendantEmail()),
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java]: Please revert this line.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandler.java [hunk 5]
```diff
@@ -33,14 +35,27 @@ public void onPaidInFullEvent(PaidInFullEvent event) {
...
+                notificationsProperties.getTemplates().getEmail().getClaimantSaysDefendantHasPaidInFull(),
+                parameters,
+                referenceForDefendant(claim.getReferenceNumber())
+            )
         );
     }
 
+    private Optional<String> getDefendantEmail(Claim claim) {
+        System.out.print("hi wtf am i here, i'm meant to be defendant: " + claim.getDefendantEmail());
+        System.out.print("THEIR DETAILS: " + claim.getClaimData().getDefendant().getEmail());
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandler.java]: Please remove above two lines. :)

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java [hunk 6]
```diff
@@ -42,17 +45,44 @@ public void setUp() {
     }
 
     @Test
-    public void sendNotificationsSendsNotificationsToDefendant() throws NotificationClientException {
-
-        PaidInFullEvent event = new PaidInFullEvent(SampleClaim.getDefault());
+    public void sendNotificationsSendsNotificationsToDefendantEmailNotLinked() throws NotificationClientException {
+        Response fullDefenceResponse = SampleResponse.FullDefence.builder().build();
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java]: We don't need `Response` as defendant is not linked yet. We just need a claim submitted by claimant here.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java [hunk 7]
```diff
@@ -42,17 +45,44 @@ public void setUp() {
...
+        PaidInFullEvent event = new PaidInFullEvent(SampleClaim.getWithResponse(fullDefenceResponse));
 
         handler.notifyDefendantForPaidInFull(event);
+        verify(notificationService, once()).sendMail(
+            eq(event.getClaim().getDefendantEmail()),
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java]: Ad defendant is not linked, this should be email provided claimant i.e. `claim.getClaimData().getDefendant().getEmail()`

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java [hunk 8]
```diff
@@ -42,17 +45,44 @@ public void setUp() {
...
         verify(notificationService, once()).sendMail(
             eq(event.getClaim().getDefendantEmail()),
             eq(CLAIMANT_SAYS_DEFENDANT_PAID_IN_FULL_TEMPLATE),
             anyMap(),
             eq(referenceForDefendant(event.getClaim().getReferenceNumber()))
         );
     }
+
+    @Test
+    public void sendNotificationsSendsNotificationsToDefendantNoEmail() throws NotificationClientException {
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/events/paindinfull/PaidInFullCitizenNotificationHandlerTest.java]: Please rename this test i.e `sendNotificationsDoesNotSendNotificationToDefendantWhenNoEmailAddress`

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandlerTest.java [hunk 9]
```diff
@@ -42,17 +43,43 @@ public void setUp() {
...
             eq(referenceForDefendant(event.getClaim().getReferenceNumber()))
         );
     }
+
+    @Test
+    public void sendNotificationsDoesNotSendNotificationToDefendantWhenNoEmailAddress() throws NotificationClientException {
+        PaidInFullEvent event = new PaidInFullEvent(SampleClaim.getDefault());
+
+        handler.notifyDefendantForPaidInFull(event);
+        verify(notificationService, never()).sendMail(
```
[the remark this answers is not in the document]
Comment (laurentaylor1) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandlerTest.java]: Gotcha thank you!

[the remark this answers is not in the document]
Comment (laurentaylor1) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandlerTest.java]: @sabahirfan can you confirm this?

[the remark this answers is not in the document]
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandlerTest.java]: agree with @andrewwa-kainos

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullStaffNotificationHandlerTest.java [hunk 10]
```diff
@@ -1,12 +1,10 @@
-package uk.gov.hmcts.cmc.claimstore.events.paindinfull;
+package uk.gov.hmcts.cmc.claimstore.events.paidinfull;
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullStaffNotificationHandlerTest.java]: 👍

Comment (laurentaylor1) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullStaffNotificationHandlerTest.java]: Lol

File: src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandlerTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (laurentaylor1) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandlerTest.java]: In order to do this, I must change
`private final String email;` to
`private String email;` in TheirDetails.java so I can set the defendant's "their details" to null. It does not permit me to set their email to null whilst building the claim.

[the remark this answers is not in the document]
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/events/paidinfull/PaidInFullCitizenNotificationHandlerTest.java]: Nope. Use SampleClaim to build a claim which does not have this email address and pass it to the event constructor.

**Pull request merged**